torch/testing/_internal/common_fsdp.py

```diff
@@ -0,0 +1,576 @@
+        end.record()
+        end.synchronize()
+        cycles_per_ms = 1000000 / start.elapsed_time(end)
+        return cycles_per_ms
+
+    # Get 10 values and remove the 2 max and 2 min and return the avg.
+    # This is to avoid system disturbance that skew the results, e.g.
+    # the very first cuda call likely does a bunch of init, which takes
+    # much longer than subsequent calls.
+    #
+    # Tested on both Tesla V100, Quadro GP100, Titan RTX, RTX 3090 GPUs
+    # and seems to return stable values. Therefore, we enable caching
+    # using lru_cache decorator above.
+    num = 10
+    vals = []
+    for _ in range(num):
+        vals.append(measure())
+    vals = sorted(vals)
+    return mean(vals[2 : num - 2])
+
+
+def dist_init(rank: int, world_size: int, filename: str, filename_rpc: str = "") -> bool:
+    """
+    Initialize torch distributed, based on a temporary file shared across ranks, which makes it possible for unrelated
+    tests to be run concurrently.
+    Return false if not enough GPUs present in the system.
+    .. warning: This limits the usecase to all ranks being on the same node
+    """
+    print(f"dist init r={rank}, world={world_size}")
+
+    os.environ["WORLD_SIZE"] = str(world_size)
+    os.environ["RANK"] = str(rank)
+    url = "file://" + filename
+
+    backend = "nccl" if torch.cuda.is_available() else "gloo"
+    if backend == "nccl" and torch.cuda.device_count() < world_size:
+        logging.warning("Requested world size cannot be reached on this machine, not enough GPUs")
+        return False
+
+    torch.distributed.init_process_group(backend=backend, rank=rank, world_size=world_size, init_method=url)
+
+    if torch.cuda.is_available() and torch.cuda.device_count():
+        torch.cuda.set_device(rank % torch.cuda.device_count())
+
+    return True
+
+
+def teardown() -> None:
+    if torch.distributed.is_initialized():
+        torch.distributed.destroy_process_group()
+
+
+def get_world_sizes() -> List[int]:
+    limit = torch.cuda.device_count()
+    return [x for x in [1, 2, 4, 8] if x <= limit]
+
+
+def rmf(filename: str) -> None:
+    """Remove a file like rm -f."""
+    try:
+        os.remove(filename)
+    except FileNotFoundError:
+        pass
+
+
+def spawn_for_all_world_sizes(test_func: Callable, world_sizes: List[int] = get_world_sizes(), args: Any = None) -> None:
+    if args is None:
+        args = []
+    for world_size in world_sizes:
+        _, filename = tempfile.mkstemp()
+        _, filename_rpc = tempfile.mkstemp()
+
+        try:
+            # (lefaudeux) Let mp handle the process joining, join=False and handling context has
+            # been unstable in the past.
+            mp.spawn(test_func, args=(world_size, filename, filename_rpc, *args), nprocs=world_size, join=True)
+        finally:
+            rmf(filename)
+            rmf(filename_rpc)
+
+
+def objects_are_equal(a: Any, b: Any, raise_exception: bool = False, dict_key: Optional[str] = None) -> bool:
```

Comment:
This function does not seem fsdp specific, so please move it to common_utils, although I'm surprised we don't have it in the codebase already.
@mruberry , are you aware of something like in common_utils?

Reply:
Of course, we have assertEqual 

https://github.com/pytorch/pytorch/blob/f9c0a39ad9320795fc8df77b570c317e0c2ab60e/torch/testing/_internal/common_utils.py#L1778

which most of the test suite uses. If we want to support stricter comparison then testing.assert_close is available:

https://pytorch.org/docs/master/testing.html?highlight=assert_close#torch.testing.assert_close

If neither of those is what we need then let me know what's missing and we should just update one or both.

Reply:
thanks, we will refactor these unit tests in follow up, see issue https://github.com/pytorch/pytorch/issues/64664